src/Storages/ObjectStorage/DataLakes/Iceberg/ManifestFile.cpp

```diff
@@ -140,6 +149,45 @@ ManifestFileContentImpl::ManifestFileContentImpl(
         content_int_column = assert_cast<const ColumnInt32 *>(content_column.get());
     }
 
+
+    Poco::JSON::Parser parser;
+
+    ColumnPtr big_partition_column = data_file_tuple_column->getColumnPtr(data_file_tuple_type.getPositionByName("partition"));
+    if (big_partition_column->getDataType() != TypeIndex::Tuple)
+    {
+        throw Exception(
+            ErrorCodes::ILLEGAL_COLUMN,
+            "The parsed column from Avro file of `partition` field should be Tuple type, got {}",
+            big_partition_column->getFamilyName());
+    }
+    const auto * big_partition_tuple = assert_cast<const ColumnTuple *>(big_partition_column.get());
+
+    auto avro_metadata = manifest_file_reader_->metadata();
+
+    std::vector<uint8_t> partition_spec_json_bytes = avro_metadata["partition-spec"];
+    String partition_spec_json_string
+        = String(reinterpret_cast<char *>(partition_spec_json_bytes.data()), partition_spec_json_bytes.size());
+    Poco::Dynamic::Var partition_spec_json = parser.parse(partition_spec_json_string);
+    const Poco::JSON::Array::Ptr & partition_specification = partition_spec_json.extract<Poco::JSON::Array::Ptr>();
+
+    std::vector<ColumnPtr> partition_columns;
+
+    for (size_t i = 0; i != partition_specification->size(); ++i)
+    {
+        auto current_field = partition_specification->getObject(static_cast<UInt32>(i));
+
+        auto source_id = current_field->getValue<Int32>("source-id");
+        PartitionTransform transform = getTransform(current_field->getValue<String>("transform"));
+
+        if (transform == PartitionTransform::Unsupported || transform == PartitionTransform::Void)
+        {
+            continue;
+        }
+
+        partition_column_infos.emplace_back(transform, source_id);
+        partition_columns.push_back(big_partition_tuple->getColumnPtr(i));
```

Comment:
I see that during schema inference we made these columns `Nullable`, but afaik elements of `partition` struct cannot contain nulls. So we can remove `Nullable` here to avoid dynamic casts later
```suggestion
        partition_columns.push_back(removeNullable(big_partition_tuple->getColumnPtr(i)));
```